File: web/src/api/pinot.ts
import { isLabel, Record } from './ipfix';

export interface PinotQueryResponse {
  resultType: string;
  resultTable?: ResultTable;
  stats: PinotStats;
  isMock: boolean;
  unixTimestamp: number;
}

export interface ResultTable {
  dataSchema: DataSchema;
  rows: string[][];
}

export interface DataSchema {
  columnNames: string[];
  columnDataTypes: string[];
}

export interface PinotStats {
  numSegmentsProcessed: number;
  numServersResponded: number;
  numSegmentsQueried: number;
  numServersQueried: number;
  numSegmentsMatched: number;
  numConsumingSegmentsQueried: number;
  numDocsScanned: number;
  numEntriesScannedInFilter: number;
  numEntriesScannedPostFilter: number;
  totalDocs: number;
  timeUsedMs: number;
  minConsumingFreshnessTimeMs: number;
  numGroupsLimitReached: boolean;
}

export function parseResultTable(resultTable?: ResultTable): Record[] {
  const result: Record[] = [];

  if (resultTable) {
    for (let i = 0; i < resultTable.rows.length; i++) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const record = { labels: {} as any, fields: {} as any };
      for (let c = 0; c < resultTable.dataSchema.columnNames.length; c++) {
        const colName = resultTable.dataSchema.columnNames[c];
        const target = isLabel(colName) ? 'labels' : 'fields';
        record[target][colName] = resultTable.rows[i][c];
      }
      result.push(record as Record);
    }
  }
  return result;
}

File: pkg/model/pinot.go
package model

import "github.com/startreedata/pinot-client-go/pinot"

// ResultType holds the type of the result
type PinotResultType string

// ResultType values
const (
	PinotResultTypeTable = "table"
)

type PinotQueryResponse struct {
	ResultType    PinotResultType    `json:"resultType"`
	ResultTable   *pinot.ResultTable `json:"resultTable,omitempty"`
	Stats         PinotStats         `json:"stats"`
	IsMock        bool               `json:"isMock"`
	UnixTimestamp int64              `json:"unixTimestamp"`
}

type PinotStats struct {
	NumSegmentsProcessed        int   `json:"numSegmentsProcessed"`
	NumServersResponded         int   `json:"numServersResponded"`
	NumSegmentsQueried          int   `json:"numSegmentsQueried"`
	NumServersQueried           int   `json:"numServersQueried"`
	NumSegmentsMatched          int   `json:"numSegmentsMatched"`
	NumConsumingSegmentsQueried int   `json:"numConsumingSegmentsQueried"`
	NumDocsScanned              int64 `json:"numDocsScanned"`
	NumEntriesScannedInFilter   int64 `json:"numEntriesScannedInFilter"`
	NumEntriesScannedPostFilter int64 `json:"numEntriesScannedPostFilter"`
	TotalDocs                   int64 `json:"totalDocs"`
	TimeUsedMs                  int   `json:"timeUsedMs"`
	MinConsumingFreshnessTimeMs int64 `json:"minConsumingFreshnessTimeMs"`
	NumGroupsLimitReached       bool  `json:"numGroupsLimitReached"`
}
